lib/exercism/authentication.rb

```diff
@@ -1,6 +1,36 @@
 class Authentication
   def self.perform(code, client_id, client_secret, target_profile)
-    id, username, email, avatar_url = Github.authenticate(code, client_id, client_secret)
-    User.from_github(id, username, email, avatar_url, target_profile)
+    new(code, client_id, client_secret, target_profile).authenticate
+  end
+
+  def initialize(code, client_id, client_secret, target_profile)
+    @code = code
+    @client_id = client_id
+    @client_secret = client_secret
+    @target_profile = target_profile
+  end
+
+  def authenticate
+    [user, access_token]
+  end
+
+  private
+
+  attr_reader :code, :client_id, :client_secret, :target_profile
+
+  def github
+    @github ||= Github.new(code, client_id, client_secret)
+  end
+
+  def access_token
+    @access_token ||= github.access_token
+  end
+
+  def user
+    User.from_github(github_user.id, github_user.username, github_user.email, github_user.avatar_url, target_profile)
```

Comment:
Great. Now to take this Introduce Parameter Object refactoring to its conclusion, you'll want to pass in the `github_user` directly, and allow `User.from_github` to extract the values as it needs them:

```ruby
  def user
    User.from_github(github_user, target_profile)
  end
```

The goal is to substitute these 4 parameters with the `GithubUser` object wherever possible.